### ISSUE TITLE
Error: Credential should be scoped to a valid region, not 'queue'.  (Aws::SQS::Errors::SignatureDoesNotMatch)

### ISSUE THREAD

**Issue by krishnan-mani** [aws/aws-sdk-ruby#1620]
### Issue description

When using the Aws::SQS::Client with credentials from a profile and the region set to "eu-west-1" to operate on a queue that is in "eu-west-1", the operation fails with this error

### Gem name ('aws-sdk', 'aws-sdk-resources' or service gems like 'aws-sdk-s3') and its version

```
$ gem list --local | grep aws
```
produces

```
aws-partitions (1.22.0)
aws-sdk-core (3.6.0)
aws-sdk-dynamodb (1.2.0)
aws-sdk-sqs (1.1.0)
aws-sigv4 (1.0.2)
```
installed using bundler with a Gemfile that looks like this

```
gem 'aws-sdk-sqs'

# Other gems omitted
```

### Version of Ruby, OS environment
ruby-2.4.0
Mac OS X Yosemite Version 10.10.5, Darwin Kernel Version 14.5.0

### Code snippets / steps to reproduce
```ruby
require 'aws-sdk-sqs'
sqs_client = Aws::SQS::Client.new(region: 'eu-west-1')
sqs_client.send_message({
  queue_url: "https://eu-west-1.queue.amazonaws.com/SOME-ACCOUNT-ID/test", 
  message_body: "hello world"
})
```

Same operation attempted with the same profile credentials and region using the aws-cli succeeds for 
version:
```
aws-cli/1.11.150 Python/2.7.10 Darwin/14.5.0 botocore/1.7.8
```

**Comment by cjyclaire** [comment]
@krishnan-mani This sounds like a credential error, how do you provide credentials, especially `region` parameter? (e.g, default credential chain? config file?) Just tested a similar case locally and cannot reproduce this. Trying to figure out behavior difference there

Another question is, for those queue urls, it usually follows pattern: https://sqs.eu-west-1.amazonaws.com/MY-ID/test1, might be a typo, yet double check : )

**Comment by cjyclaire** [comment]
Closing due to ages, yet feel free to reopen with further info

**Comment by Nomane** [comment]
Issue is located in SQS endpoint returned by AWS API.

In my case I use aws cli by this way

> aws sqs get-queue-url --queue-name ${context.queueName}

which return me something like this:

> https://eu-central-1.queue.amazonaws.com/XXXXXXXX/myqueuename

It's not a valid queue URL, you have to use something like

> https://sqs.eu-central-1.amazonaws.com/XXXXXXXX/myqueuename

To fix this I force the endpoint like that:

> aws sqs get-queue-url --queue-name ${context.queueName} **--endpoint-url** https://sqs.eu-central-1.amazonaws.com/**

Hope that could be useful

**Comment by immerrr** [comment]
When you're saying "not a valid queue URL", you must have some grounds to believe that. I have no problem accessing queues with `https://eu-west-1.queue.amazonaws.com` endpoint via boto3 library, hence I believe the problem is in ruby sdk.

The error message says that "queue" is not a valid region for url that looks like
```
https://eu-west-1.queue.amazonaws.com
```
but works fine for URLs that look like
```
https://sqs.eu-west-1.amazonaws.com
```
I'd suggest that queue region is extracted from the second dot-separated component of endpoint netloc, which is an incorrect assumption in case of `{REGION}.queue.*` URLs. And it took me 3mins from that point to finding [a place that looks suspiciously close](https://github.com/aws/aws-sdk-ruby/blob/9ea447b7e9e57e27bb7924aca40a5bf0c3265a94/gems/aws-sdk-sqs/lib/aws-sdk-sqs/plugins/queue_urls.rb#L25) to what I'm describing (disclaimer: i have no experience with ruby, so I might be wrong here).

Pinging @cjyclaire.

**Comment by immerrr** [comment]
Ok, according to https://docs.aws.amazon.com/general/latest/gr/rande.html#sqs_region, `sqs.{REGION}` URLs are the new and correct ones, and `{REGION}.queue` ones are legacy that are somehow permitted "If you use the AWS CLI or SDK for Python". Still, we've faced this problem today connecting Ruby & Python code, so I strongly believe this is worth fixing.

**Comment by cjyclaire** [comment]
@immerrr I tried `#get_queue_url` API, and url response is following pattern
`https://sqs.us-west-2.amazonaws.com/123.../foo....`, Ruby SDK doesn't have any customization around the response url strings SQS send back.

As you mentioned, the CLI pattern is likely a customization on their side or a legacy pattern, I'm not super familiar with boto/CLI code base, I'll need take a deeper look to find out. For the code link you mentioned, it's helping for signning regions, doesn't change url provided :)

**Comment by immerrr** [comment]
@cjyclaire API returns you a queue URL based on the endpoint you use:

```
In [10]:  boto3.client('sqs', endpoint_url='https://eu-west-1.queue.amazonaws.com').get_queue_url(QueueName='hello')['QueueUrl']
Out[10]: 'https://eu-west-1.queue.amazonaws.com/{ACCOUNT_ID}/hello'

In [11]:  boto3.client('sqs', endpoint_url='https://sqs.eu-west-1.amazonaws.com').get_queue_url(QueueName='hello')['QueueUrl']
Out[11]: 'https://sqs.eu-west-1.amazonaws.com/{ACCOUNT_ID}/hello'
```

**Comment by immerrr** [comment]
> For the code link you mentioned, it's helping for signning regions, doesn't change url provided :)

I'm not saying it changes the URL, I'm saying it takes 2nd dot-separated netloc component as region name, and uses it for signing, and naturally fails when that 2nd dot-separated netloc component is `queue` as in the legacy patterns rather than a region name.

**Comment by cjyclaire** [comment]
>I'm not saying it changes the URL, I'm saying it takes 2nd dot-separated netloc component as region name, and uses it for signing, and naturally fails when that 2nd dot-separated netloc component is queue as in the legacy patterns rather than a region name.

Correct, what I'm trying to figure out is whether this is an old pattern that AWS server send back original or it's simply a boto customization. If it's simply a boto only thing, it's not appropriate for Ruby to adding the hack simply for in sync with boto. If it's a server raw response that regressed, we might consider adding a work around for it.

**Comment by awood45** [comment]
I'm seeing where the issue is here. This code reproduces the issue from scratch:

```ruby
sqs = Aws::SQS::Client.new(
  region: "eu-west-1",
  endpoint: "https://eu-west-1.queue.amazonaws.com"
)
queue_url = sqs.get_queue_url(queue_name: QUEUE_NAME)
sqs.send_message(
  queue_url: queue_url,
  message_body: "Test"
)
```

It is true that, unless you're intentionally redirecting towards the legacy endpoint, you're not going to run into this issue unless copying the queue URL from the Python/CLI result when also choosing the legacy endpoint.

This would be a feature request to add the same sort of customizations to the Ruby SDK. There is probably a safe way to do it, but it would also not likely be at the top of the feature request backlog. I am open to reviewing a PR to add support for this.

**Comment by immerrr** [comment]
> unless copying the queue URL from the Python/CLI

This is exactly the way I've encountered the problem: when copying a queue URL from Python part of the infra to Ruby. Also, FTR, it was not _a choice_ to use the legacy endpoint, it is the default at least as of versions
```
In [19]: boto3.__version__
Out[19]: '1.4.4'

In [21]: botocore.__version__
Out[21]: '1.5.95'
```
The former was released in Jan 2017, the latter -- in Aug 2017, so the libraries themselves are not exactly legacy.

**Comment by awood45** [comment]
Now that is definitely interesting. I'll investigate what's going on with this.

**Comment by immerrr** [comment]
Apparently, that is still the default endpoint: https://github.com/boto/botocore/blob/develop/botocore/data/endpoints.json#L2232

**Comment by immerrr** [comment]
As per the developers, the old endpoints are used because Py2.6 is unable to work with "subjectAltName" extension, so it will not be fixed until 2.6 is dropped, which should be soon, but unclear yet.
